### PR TITLE
Revamp bugsnag-android docs

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Revamps the bugsnag-android docs so that they are a bit easier to follow and cover more topics.

## Changeset

- Moved docs intended mainly for internal consumption to a separate `docs` folder, which is linked to from the main contributing guide/readme
- Removed obsolete Travis CI badge
- Added further detail on code quality tools which have been added recently
- Removed bintray from release instructions
- Added separate sections for several key concepts:
  - An environment setup page which lists the pre-requisite programs that are required to develop bugsnag-android
  - A project structure page to give a high level overview of how the bugsnag-android project works
  - An easily accessible index page which links into all the disparate docs pages
  - A separate page for mazerunner E2E tests

### CODE REVIEW

File: docs/DEVELOPMENT.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+Please see the [mazerunner docs](MAZERUNNER.md) for information on how to run E2E tests locally.
+
+## Example app
+
+The example app can be found in [examples/sdk-app-example](../examples/sdk-app-example). It contains example code which triggers crashes.
+
+You should open `examples/sdk-app-example` as a separate Android Studio project and [run the app](https://developer.android.com/training/basics/firstapp/running-app). This will require a connected device or emulator.
+
+The app **does not use the local development version of Bugsnag**. To test local development changes, you will need to build the library locally with a version number of `9.9.9`, and update the `bugsnag-android` dependency in `examples/sdk-app-example/build.gradle` to use that version.
```
Comment (twometresteve) [inline docs/DEVELOPMENT.md:24]:
```suggestion
The app **does not use the local development version of Bugsnag**. To test local development changes, you will need to build the library locally with an arbitrary version number, e.g. `9.9.9`, and update the `bugsnag-android` dependency in `examples/sdk-app-example/build.gradle` to use that version.
```

File: docs/MAZERUNNER.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Mazerunner tests
+
+E2E tests are implemented with our notifier testing tool [Maze runner](https://github.com/bugsnag/maze-runner), which is a black-box test framework written in Ruby.
+
+## Running remote end-to-end tests
+
+End to end tests are written in cucumber-style `.feature` files, and need Ruby-backed "steps" in order to know what to run. The tests are located in the top level [`features`](/features/) directory.
+
+Maze Runner's CLI and the test fixtures are containerized so you'll need Docker (and Docker Compose) to run them.
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:9]: A docker image is provided for Maze Runner, but you don't have to use it and have docker/compose installed.  You can quite happily run it as a Gem (and indeed that is what the later instruction here do).  So I think you can simply remove this line, or just note that Ruby and Bundler are required.

File: docs/MAZERUNNER.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+__Note: only Bugsnag employees can run the end-to-end tests.__ We have dedicated test infrastructure and private BrowserStack credentials which can't be shared outside of the organisation.
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:11]: Use US spelling
```suggestion
__Note: only Bugsnag employees can run the end-to-end tests.__ We have dedicated test infrastructure and private BrowserStack credentials which can't be shared outside of the organization.
```

File: docs/MAZERUNNER.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+aws configure --profile=opensource
+```
+
+Subsequently you'll need to run the following commmand to authenticate with the registry:
+
+```
+$(aws ecr get-login --profile=opensource --no-include-email)
+```
+
+__Your session will periodically expire__, so you'll need to run this command to re-authenticate when that happens.
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:27]: None of this is needed any more.

File: docs/MAZERUNNER.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+Remote tests can be run against real devices provided by BrowserStack. In order to run these tests, you need to set the following environment variables:
+
+- A BrowserStack App Automate Username: `MAZE_DEVICE_FARM_USERNAME`
+- A BrowserStack App Automate Access Key: `MAZE_DEVICE_FARM_ACCESS_KEY`
+- A path to a [BrowserStack local testing binary](https://www.browserstack.com/local-testing/app-automate): `MAZE_BS_LOCAL`
+
+### Running an end-to-end test
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:35]: Move this up ~7 lines to before the previous section.

File: docs/MAZERUNNER.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+1. Build the test fixtures `make test-fixtures` (separate fixtures are built with and without the NDK/ANR plugins)
+1. Check the contents of `Gemfile` to select the version of `maze-runner` to use
+1. To run a single feature:
+    ```shell script
+    make test-fixtures && \
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:41]: This duplicates the first step.

File: docs/MAZERUNNER.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+- A BrowserStack App Automate Username: `MAZE_DEVICE_FARM_USERNAME`
+- A BrowserStack App Automate Access Key: `MAZE_DEVICE_FARM_ACCESS_KEY`
+- A path to a [BrowserStack local testing binary](https://www.browserstack.com/local-testing/app-automate): `MAZE_BS_LOCAL`
+
+### Running an end-to-end test
+
+1. Build the test fixtures `make test-fixtures` (separate fixtures are built with and without the NDK/ANR plugins)
+1. Check the contents of `Gemfile` to select the version of `maze-runner` to use
+1. To run a single feature:
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:39]: Probably need a step to run `bundle install` before the first use.

File: docs/MAZERUNNER.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    ```shell script
+    make test-fixtures && \
+    bundle exec maze-runner --app=build/fixture.apk                 \
+                            --farm=bs                               \
+                            --device=ANDROID_9_0                    \
+                            features/smoke_tests/unhandled.feature
+    ```
+1. To run all features, omit the final argument.
```
Comment (twometresteve) [inline docs/MAZERUNNER.md:47]:
```suggestion
1. To run all features, omit the final argument, but be wary of how many tests you run locally as we have a limited number of parallel tests and local running subverts the controls we have in place.  For a full test run it is generally best to push your branch to Github and let CI run them.
```

File: docs/PROJECT_STRUCTURE.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+## Modules
+
+The project consists of [6 Gradle modules](https://gradle.org/), which are linked to below:
+
+- [bugsnag-android-core](../bugsnag-android-core/README.md) - contains the core functionality required to capture and deliver JVM errors to Bugsnag.
+- [bugsnag-plugin-android-ndk](../bugsnag-plugin-android-ndk/README.md) - contains optional functionality that installs signal handlers and captures NDK errors.
+- [bugsnag-plugin-android-anr](../bugsnag-plugin-android-ndk/README.md) - contains optional functionality that installs a signal handler to capture ANR errors.
+- [bugsnag-plugin-react-native](../bugsnag-plugin-react-native/README.md) - contains optional functionality that serializes information into a format that can be understood by the React Native bridge.
+- [bugsnag-android](../bugsnag-android/README.md) - an anchor package which allows users to pull in all the required modules.
+- [bugsnag-android-ndk](../bugsnag-android-ndk/README.md) - an anchor package which allows users to pull in all the required modules. Published for legacy reasons
```
Comment (twometresteve) [inline docs/PROJECT_STRUCTURE.md:14]: nit:
```suggestion
- [bugsnag-android-ndk](../bugsnag-android-ndk/README.md) - an anchor package which allows users to pull in all the required modules. Published for legacy reasons.
```

File: docs/PROJECT_STRUCTURE.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+The project is containerized with [dockerfiles](../dockerfiles). These are used to test the project on CI with [Buildkite](http://buildkite.com/).
+
+## Git submodule
+
+The project uses one git submodule to access [libunwindstack](https://github.com/bugsnag/libunwindstack-ndk), which is used for capturing NDK stacktraces.
+
+## E2E tests
+
+Bugsnag makes extensive use of E2E testing with [mazerunner](https://github.com/bugsnag/maze-runner), which is our customer black-box testing framework written in Ruby.
```
Comment (twometresteve) [inline docs/PROJECT_STRUCTURE.md:30]: Customer?  Should that be "custom", or just not there at all?

```suggestion
Bugsnag makes extensive use of E2E testing with [mazerunner](https://github.com/bugsnag/maze-runner), which is our custom black-box testing framework written in Ruby.
```

File: docs/RELEASING.md [hunk 11]
```diff
@@ -1,41 +1,10 @@
 # Releasing a New Version
 
-If you are a project maintainer, you can build and release a new version of
-`bugsnag-android` as follows:
+`bugsnag-android` is released via [Sonatype](https://oss.sonatype.org/). If you are a project maintainer you can release a new version by unblocking the publish step on CI and following the steps below.
 
-## One-time setup
+## Pre-release Checklist
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply docs/RELEASING.md:5]: tweaked to sentence case

File: docs/MAZERUNNER.md (hunk 7 above)
Comment (fractalwrench) [reply docs/MAZERUNNER.md:39]: Added a note.

File: docs/MAZERUNNER.md (hunk 4 above)
Comment (fractalwrench) [reply docs/MAZERUNNER.md:27]: Removed these lines.

File: docs/MAZERUNNER.md (hunk 2 above)
Comment (fractalwrench) [reply docs/MAZERUNNER.md:9]: Removed this line.

File: docs/MAZERUNNER.md (hunk 6 above)
Comment (fractalwrench) [reply docs/MAZERUNNER.md:41]: I always forget to run `make test-fixtures` and get confused with the results. I've therefore removed the first step and kept the command as-is

File: docs/MAZERUNNER.md (hunk 5 above)
Comment (twometresteve) [reply docs/MAZERUNNER.md:17]: Retracted - my mistake.

File: docs/MAZERUNNER.md (hunk 6 above)
Comment (twometresteve) [reply docs/MAZERUNNER.md:23]: I guess that's fine, it's just potentially a waste of time for people who run it without thinking about what it's actually doing.  If you're just running different tests without changing the notifier or scenario classes then you don't need to rebuild it - but I can live with it as it is now!

**Review by twometresteve — APPROVED** [review]
LGTM

**Pull request merged**